File: org-app/internal/db/tenant_registry.go
package db

import (
	"database/sql"
	"fmt"
	"log"

	_ "github.com/lib/pq"
)

const registrySchema = `
CREATE TABLE IF NOT EXISTS tenants (
	id   TEXT PRIMARY KEY,
	host TEXT NOT NULL
);
`

func OpenTenantRegistry(conn string) (*sql.DB, error) {
	db, err := sql.Open("postgres", conn)
	if err != nil {
		log.Printf("Unable to open connection to DB: %s", conn)
		return nil, err
	}
	if err := db.Ping(); err != nil {
		log.Printf("Unable to ping DB: %s", conn)
		return nil, err
	}
	if _, err := db.Exec(registrySchema); err != nil {
		log.Printf("Unable to create registry table")
		return nil, fmt.Errorf("registry schema: %w", err)
	}
	return db, nil
}

func RegisterTenant(db *sql.DB, tenantID, host string) error {
	_, err := db.Exec(
		`INSERT INTO tenants (id, host) VALUES ($1, $2)`,
		tenantID, host)

	if err != nil {
		log.Printf("failed to insert tenant into tenant table: %s", tenantID)
	}
	log.Printf("inserted tenant into tenant table: %s", tenantID)
	return err
}

func GetTenantHost(db *sql.DB, tenantID string) (host string, err error) {
	err = db.QueryRow("SELECT host FROM tenants WHERE id = $1", tenantID).Scan(&host)
	return host, err
}

type Registry struct {
	db *sql.DB
}

func NewRegistry(db *sql.DB) *Registry {
	return &Registry{db: db}
}

func (r *Registry) Register(tenantID, host string) error {
	return RegisterTenant(r.db, tenantID, host)
}

func (r *Registry) Host(tenantID string) (string, error) {
	return GetTenantHost(r.db, tenantID)
}

File: todo-app/lib/todo_app/tenant_registry.ex
defmodule TodoApp.TenantRegistry do
  @moduledoc "TenantDB: registry of tenant_id -> host. DB name for a tenant = todo-app_<tenant_id>"
  use GenServer
  require Logger

  def start_link(opts \\ []) do
    GenServer.start_link(__MODULE__, opts, name: __MODULE__)
  end

  @impl true
  def init(_opts) do
    cfg = Application.get_env(:todo_app, :tenant_registry)
    db_opts = [
      hostname: cfg[:hostname],
      port: cfg[:port],
      username: cfg[:username],
      password: cfg[:password],
      database: cfg[:database]
    ]
    {:ok, pid} = Postgrex.start_link(db_opts)
    create_schema(pid)
    {:ok, %{conn: pid, cfg: cfg}}
  end

  defp create_schema(pid) do
    Postgrex.query!(pid, """
    CREATE TABLE IF NOT EXISTS tenants (
      id   TEXT PRIMARY KEY,
      host TEXT NOT NULL
    )
    """, [])
  end

  def register(tenant_id, host) do
    GenServer.call(__MODULE__, {:register, tenant_id, host})
  end

  def get_host(tenant_id) do
    GenServer.call(__MODULE__, {:get_host, tenant_id})
  end

  @impl true
  def handle_call({:register, tenant_id, host}, _from, state) do
    res = Postgrex.query(state.conn, "INSERT INTO tenants (id, host) VALUES ($1, $2) ON CONFLICT (id) DO UPDATE SET host = EXCLUDED.host", [tenant_id, host])
    {:reply, res, state}
  end

  def handle_call({:get_host, tenant_id}, _from, state) do
    case Postgrex.query(state.conn, "SELECT host FROM tenants WHERE id = $1", [tenant_id]) do
      {:ok, %{rows: [[host]]}} -> {:reply, {:ok, host}, state}
      _ -> {:reply, :error, state}
    end
  end
end
